File: packages/slim_report_core/src/slim_report_core/rendering/qrcode.py
"""Deterministic QR-style matrix helpers."""

from __future__ import annotations

QR_GRID_SIZE = 29
QR_QUIET_ZONE = 4
QR_DATA_SIZE = 21


def qr_module_matrix(value: object) -> list[list[bool]]:
    """Return a stable QR-like matrix with quiet zone and finder patterns.

    This is a dependency-free visual fallback, not a standards-compliant QR encoder.
    """

    text = str(value or "QR Code")
    seed = _stable_hash(text)
    matrix = [[False for _ in range(QR_GRID_SIZE)] for _ in range(QR_GRID_SIZE)]

    finder_origins = [
        (QR_QUIET_ZONE, QR_QUIET_ZONE),
        (QR_QUIET_ZONE, QR_QUIET_ZONE + QR_DATA_SIZE - 7),
        (QR_QUIET_ZONE + QR_DATA_SIZE - 7, QR_QUIET_ZONE),
    ]
    for row, col in finder_origins:
        _draw_finder(matrix, row, col)

    timing_row = QR_QUIET_ZONE + 6
    timing_col = QR_QUIET_ZONE + 6
    for index in range(8, QR_DATA_SIZE - 8):
        enabled = index % 2 == 0
        matrix[timing_row][QR_QUIET_ZONE + index] = enabled
        matrix[QR_QUIET_ZONE + index][timing_col] = enabled

    for inner_row in range(QR_DATA_SIZE):
        for inner_col in range(QR_DATA_SIZE):
            row = QR_QUIET_ZONE + inner_row
            col = QR_QUIET_ZONE + inner_col
            if _reserved(inner_row, inner_col):
                continue
            char_code = ord(text[(inner_row * QR_DATA_SIZE + inner_col) % len(text)])
            mixed = (
                seed
                ^ ((inner_row + 1) * 0x45D9F3B)
                ^ ((inner_col + 1) * 0x27D4EB2D)
                ^ (char_code * (inner_row + inner_col + 1))
            ) & 0xFFFFFFFF
            matrix[row][col] = mixed % 7 in {0, 2, 5}

    return matrix


def _stable_hash(value: str) -> int:
    seed = 2166136261
    for char in value:
        seed ^= ord(char)
        seed = (seed * 16777619) & 0xFFFFFFFF
    return seed


def _draw_finder(matrix: list[list[bool]], top: int, left: int) -> None:
    for row in range(7):
        for col in range(7):
            edge = row in {0, 6} or col in {0, 6}
            core = 2 <= row <= 4 and 2 <= col <= 4
            matrix[top + row][left + col] = edge or core


def _reserved(row: int, col: int) -> bool:
    finder_origins = [(0, 0), (0, QR_DATA_SIZE - 7), (QR_DATA_SIZE - 7, 0)]
    for finder_row, finder_col in finder_origins:
        if finder_row - 1 <= row <= finder_row + 7 and finder_col - 1 <= col <= finder_col + 7:
            return True
    return row == 6 or col == 6

File: packages/slim_report_designer_ui/slim_report_designer_ui/static/js/qrcode.js
const QR_GRID_SIZE = 29;
const QR_QUIET_ZONE = 4;
const QR_DATA_SIZE = 21;

export function appendQrSvg(container, {
  value = "QR Code",
  left = 0,
  top = 0,
  size = 0,
  foreground = "currentColor",
  background = "#ffffff"
} = {}) {
  const svg = document.createElementNS("http://www.w3.org/2000/svg", "svg");
  svg.classList.add("qrcode-grid");
  svg.setAttribute("viewBox", `0 0 ${QR_GRID_SIZE} ${QR_GRID_SIZE}`);
  svg.setAttribute("role", "img");
  svg.setAttribute("aria-label", String(value || "QR Code"));
  svg.setAttribute("shape-rendering", "crispEdges");
  svg.style.left = `${left}px`;
  svg.style.top = `${top}px`;
  svg.style.width = `${size}px`;
  svg.style.height = `${size}px`;

  if (!isTransparent(background)) {
    const backgroundRect = document.createElementNS("http://www.w3.org/2000/svg", "rect");
    backgroundRect.setAttribute("width", String(QR_GRID_SIZE));
    backgroundRect.setAttribute("height", String(QR_GRID_SIZE));
    backgroundRect.setAttribute("fill", background);
    svg.appendChild(backgroundRect);
  }

  const group = document.createElementNS("http://www.w3.org/2000/svg", "g");
  group.setAttribute("fill", foreground);
  for (const [rowIndex, row] of qrModuleMatrix(value).entries()) {
    for (const [colIndex, enabled] of row.entries()) {
      if (!enabled) {
        continue;
      }
      const cell = document.createElementNS("http://www.w3.org/2000/svg", "rect");
      cell.setAttribute("x", String(colIndex));
      cell.setAttribute("y", String(rowIndex));
      cell.setAttribute("width", "1");
      cell.setAttribute("height", "1");
      group.appendChild(cell);
    }
  }
  svg.appendChild(group);
  container.appendChild(svg);
  return svg;
}

export function qrSvgMarkup(value = "QR Code", {
  foreground = "#111827",
  background = "#ffffff"
} = {}) {
  const safeValue = escapeHtml(value || "QR Code");
  const backgroundRect = isTransparent(background)
    ? ""
    : `<rect width="${QR_GRID_SIZE}" height="${QR_GRID_SIZE}" fill="${escapeHtml(background)}" />`;
  const cells = [];
  for (const [rowIndex, row] of qrModuleMatrix(value).entries()) {
    for (const [colIndex, enabled] of row.entries()) {
      if (enabled) {
        cells.push(`<rect x="${colIndex}" y="${rowIndex}" width="1" height="1" />`);
      }
    }
  }
  return `<svg role="img" aria-label="${safeValue}" viewBox="0 0 ${QR_GRID_SIZE} ${QR_GRID_SIZE}" xmlns="http://www.w3.org/2000/svg" shape-rendering="crispEdges" style="display:block;width:100%;height:100%">${backgroundRect}<g fill="${escapeHtml(foreground)}">${cells.join("")}</g></svg>`;
}

export function qrModuleMatrix(value = "QR Code") {
  const text = String(value || "QR Code");
  const seed = stableHash(text);
  const matrix = Array.from({ length: QR_GRID_SIZE }, () => Array(QR_GRID_SIZE).fill(false));
  const finderOrigins = [
    [QR_QUIET_ZONE, QR_QUIET_ZONE],
    [QR_QUIET_ZONE, QR_QUIET_ZONE + QR_DATA_SIZE - 7],
    [QR_QUIET_ZONE + QR_DATA_SIZE - 7, QR_QUIET_ZONE]
  ];

  for (const [row, col] of finderOrigins) {
    drawFinder(matrix, row, col);
  }

  const timingRow = QR_QUIET_ZONE + 6;
  const timingCol = QR_QUIET_ZONE + 6;
  for (let index = 8; index < QR_DATA_SIZE - 8; index += 1) {
    const enabled = index % 2 === 0;
    matrix[timingRow][QR_QUIET_ZONE + index] = enabled;
    matrix[QR_QUIET_ZONE + index][timingCol] = enabled;
  }

  for (let innerRow = 0; innerRow < QR_DATA_SIZE; innerRow += 1) {
    for (let innerCol = 0; innerCol < QR_DATA_SIZE; innerCol += 1) {
      if (reserved(innerRow, innerCol)) {
        continue;
      }
      const row = QR_QUIET_ZONE + innerRow;
      const col = QR_QUIET_ZONE + innerCol;
      const charCode = text.charCodeAt((innerRow * QR_DATA_SIZE + innerCol) % text.length);
      const mixed = (
        seed
        ^ Math.imul(innerRow + 1, 0x45d9f3b)
        ^ Math.imul(innerCol + 1, 0x27d4eb2d)
        ^ Math.imul(charCode, innerRow + innerCol + 1)
      ) >>> 0;
      matrix[row][col] = [0, 2, 5].includes(mixed % 7);
    }
  }

  return matrix;
}

function drawFinder(matrix, top, left) {
  for (let row = 0; row < 7; row += 1) {
    for (let col = 0; col < 7; col += 1) {
      const edge = row === 0 || row === 6 || col === 0 || col === 6;
      const core = row >= 2 && row <= 4 && col >= 2 && col <= 4;
      matrix[top + row][left + col] = edge || core;
    }
  }
}

function reserved(row, col) {
  const finderOrigins = [
    [0, 0],
    [0, QR_DATA_SIZE - 7],
    [QR_DATA_SIZE - 7, 0]
  ];
  return finderOrigins.some(([finderRow, finderCol]) => (
    row >= finderRow - 1
    && row <= finderRow + 7
    && col >= finderCol - 1
    && col <= finderCol + 7
  )) || row === 6 || col === 6;
}

function stableHash(value) {
  let seed = 2166136261;
  for (let index = 0; index < value.length; index += 1) {
    seed ^= value.charCodeAt(index);
    seed = Math.imul(seed, 16777619) >>> 0;
  }
  return seed;
}

function isTransparent(value) {
  return ["", "none", "transparent"].includes(String(value || "").trim().toLowerCase());
}

function escapeHtml(value) {
  return String(value).replace(/[&<>"']/g, (char) => ({
    "&": "&amp;",
    "<": "&lt;",
    ">": "&gt;",
    "\"": "&quot;",
    "'": "&#39;"
  })[char]);
}
